lib/transactions/transaction.ts

```diff
@@ -0,0 +1,228 @@
+            const result: ITransactionBroadcasted = await this.broadCastTransactionFromHex(this.signedTransactionBytes, remote)
+            this.#onBroadcasted(result);
+            return result;
+        } else {
+            throw new Error('broadCastTransaction cannot be processed. transaction=' + JSON.stringify(this));
+        }
+    }
+
+    async broadCastTransactionFromHex(signedTransactionHex: string, remote: RemoteAPICaller): Promise<ITransactionBroadcasted> {
+        const data = await remote.apiCall('post', { requestType: 'broadcastTransaction', transactionBytes: signedTransactionHex });
+        return data as unknown as ITransactionBroadcasted;
+    }
+
+    canBroadcast(): boolean {
+        return Converters.isHex(this.signedTransactionBytes) && this.state === TransactionState.SIGNED;
+    }
+
+    #onBroadcasted(result: ITransactionBroadcasted) {
+        if (this.canBroadcast()) {
+            this._transactionID = result.transaction;
+            this._fullHash = result.fullHash;
+            this._state = TransactionState.BROADCASTED;
+        } else {
+            throw new Error('Something went wrong on transaction broadcast');
+        }
+
+    }
+    //=======================================
+
+    ////// getters
+    get requestJSON(): IRequestJSON | null {
+        return this._requestJSON;
+    }
+
+    get state(): TransactionState {
+        return this._state;
+    }
+
+    get unsignedTransactionBytes(): string | null {
+        return this._unsignedTransactionBytes;
+    }
+
+    get signedTransactionBytes(): string | null {
+        return this._signedTransactionBytes;
+    }
+
+    //static functions
+    static createTransactionFromRequestJSON(requestJSON: IRequestJSON) {
+        return new Transaction(requestJSON);
+    }
+
+    static createTransactionFromBytes(bytes: string, signed: boolean) {
+        const transaction = new Transaction();
+        if (signed) {
+            transaction._signedTransactionBytes = bytes;
+            transaction._state = TransactionState.SIGNED;
+        } else {
+            transaction._unsignedTransactionBytes = bytes;
+            transaction._state = TransactionState.UNSIGNED;
+        }
+    }
+
+    static async getTransactionJSONFromBytes(bytes: string, remote: RemoteAPICaller) {
+        const data = await remote.apiCall('get', { requestType: 'parseTransaction', transactionBytes: bytes });
+        return data as unknown as ITransactionJSON;
+    }
+}
+
+
+export interface IRequestJSON {
+    requestType: string,
+    recipient: string,
+    publicKey: string,
+    feeNQT: string,
+    deadline: number,
+    message?: string
+}
+
+export interface ITransactionBroadcasted {
+    fullHash: string,
+    transaction: string
+}
+
+export interface ITransaction {
+    transactionJSON: ITransactionJSON,
+}
+
+export interface ITransactionJSON {
+    senderPublicKey: string,
+    feeNQT: string,
+    type: number,

```

Comment:
Do you have exact types here?

Reply:
Yes, "type" is a number 0 to 7 representing in order the following types of transactions: {"payment", "messaging", "colored_coins", "digital_goods", "account_control", "monetary_system", "data", "shuffling"};
These transaction types are defined in in the Coop network node, and are always returned as a number (not string).

Reply:
By the way, we also have transaction subtype which is still a number, but it's value depends on "type". These are the subtypes (each of the 8 rows corresponds to a 'type' respectively), and the subtype is the inferred index in the respective array.
```
subtype = {
            {"ordinary_payment"},
            {"arbitrary_message", "alias_assignment", "poll_creation", "vote_casting", "hub_announcement", "account_info", "alias_sell", "alias_buy", "alias_delete", "phasing_vote_casting", "account_property", "account_property_delete"},
            {"asset_issuance", "asset_transfer", "ask_order_placement", "bid_order_placement", "ask_order_cancellation", "bid_order_cancellation", "dividend_payment", "asset_delete", "asset_increase", "property_set", "property_delete"},
            {"listing", "delisting", "price_change", "quantity_change", "purchase", "delivery", "feedback", "refund"},
            {"effective_balance_leasing", "phasing_only"},
            {"currency_issuance", "reserve_increase", "reserve_claim", "currency_transfer", "publish_exchange_offer", "exchange_buy", "exchange_sell", "currency_minting", "currency_deletion"},
            {"upload", "extend"},
            {"creation", "registration", "processing", "recipients", "verification", "cancellation"}
    };
```